tools/c7n_azure/c7n_azure/actions/firewall.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2019 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+
+from c7n_azure.actions.base import AzureBaseAction
+from c7n_azure.utils import resolve_service_tag_alias
+from netaddr import IPAddress
+
+from c7n.filters.core import type_schema
+
+
+class SetFirewallAction(AzureBaseAction):
+
+    schema = type_schema(
+        'set-firewall-rules',
+        required=[],
+        **{
+            'append': {'type': 'boolean', 'default': False},
+            'bypass-rules': {'type': 'array'},
+            'ip-rules': {'type': 'array', 'items': {'type': 'string'}},
+            'virtual-network-rules': {'type': 'array', 'items': {'type': 'string'}}
+        }
+    )
+
+    @abstractmethod
+    def __init__(self, data, manager=None):
+        super(SetFirewallAction, self).__init__(data, manager)
+
+    def _prepare_processing(self):
+        self.client = self.manager.get_client()
+        self.append = self.data.get('append', False)
+
+    @abstractmethod
+    def _process_resource(self, resource):
+        pass
+
+    def _build_bypass_rules(self, existing_bypass, new_rules):
+        if self.append:
+            without_duplicates = [r for r in existing_bypass if r not in new_rules]
+            new_rules.extend(without_duplicates)
+        return ','.join(new_rules or ['None'])
+
+    def _build_vnet_rules(self, existing_vnet, new_rules):
+        if self.append:
+            without_duplicates = [r for r in existing_vnet if r not in new_rules]
+            new_rules.extend(without_duplicates)
+        return new_rules
+
+    def _build_ip_rules(self, existing_ip, new_rules):
+        rules = []
+        for rule in new_rules:
+            resolved_set = resolve_service_tag_alias(rule)
+            if resolved_set:
```

Comment:
Can this `if` statement somehow be moved to its own method and explain what it is checking/doing? I'm not following what and why you are doing all these different lists/ range/ len checks. A separate method with a descriptive name would help.

Reply:
I went ahead and added extensive comments in there.

Reply:
I went ahead and added extensive comments in there, although for some reason in my github view they aren't showing as outdated/updated in this comment, seems locked to a commit :(